### PR TITLE
BUG: Clear error before and after PyErr_WarnEx for 3118 buffers

### LINKED ISSUE #6741: ctypeslib.as_array aborts
This script fails:

```
import numpy as np
import ctypes

sh = (1, 1)
tmp = np.zeros(sh, np.uint8)
test = ctypes.cast(ctypes.c_void_p(tmp.__array_interface__['data'][0]), ctypes.POINTER(ctypes.c_ubyte))
np.ctypeslib.as_array(test, sh)
```

python 3.4.2-dbg
numpy==1.10.1
Ubuntu 14.10
1. Under release python this code runs without crashes, because of disabled assertions
2. Under debug release python asserts failed:
   `python3.4: ../Objects/abstract.c:2077: PyObject_Call: Assertion '(result != ((void *)0) && !PyErr_Occurred()) || (result == ((void *)0) && PyErr_Occurred())' failed.`

Can you please explain why?

P.S. may be it can help:
If `sh = (1, 1)` change to `sh = 239`:
1. Release python will raise `TypeError: shape must be a tuple`
2. Debug python will also aborts

### PR DESCRIPTION
**Pull request by tynn**
Fixes #6741

The `PyErr_WarnEx` seems to be the reason of the `PyObject_Call: Assertion` failure.

Here I assumed that `PyErr_WarnEx` is primarily used to print the warning message and suppress all errors. Thus `PyErr_Clear` should be safe to be used to clear the error which is turned into the warning and clear all errors possibly set by the warning.

From [PyErr_WarnEx](https://docs.python.org/3/c-api/exceptions.html#issuing-warnings)
> They normally print a warning message to sys.stderr; however, it is also possible that the user has specified that warnings are to be turned into errors, and in that case they will raise an exception. It is also possible that the functions raise an exception because of a problem with the warning machinery.

The tests relate to #6214

### CODE REVIEW

**Comment by njsmith** [comment]
From a quick glance at the patch, the use of `PyErr_Clear` here seems like it must be wrong. If something is setting an error then we almost always should exit and propagate the exception, not ignore the error and continue. And in the rare cases where `PyErr_Clear`is the right thing to use, it should be called immediately after the potentially-failing operation, with a comment explaining why we want to ignore errors there.

**Comment by tynn** [comment]
The usage of `PyErr_WarnEx` seems wrong here. Also `_array_from_buffer_3118` seems to not follow some Python conventions. It returns `-1`, but doesn't set an error. With the usage of this function, errors are ignored anyway. And I'm not sure if the warnings should ever make the function fail.

The issue itself arises after ` _descriptor_from_pep3118_format` fails and sets an error which was never handled. Would you accept it if `PyErr_Clear` is called just before `PyBytes_FromFormat`, or would you prefer rewriting this part completely?

**Comment by seberg** [comment]
Why does it give a warning and not just error out at all? Not following python conventions is maybe not nice, but not so bad for an internal function, but I think the problem here may be that the function should be able to signal three and not two possible results:
  1. Everything is good
  2. This is not a buffer at all (some error may have happened, but we should have cleared it)
  3. It seems to be an invalid buffer

And it seems for 3. it actually sets a warning in a pretty nasty way, since that warning can also easily be an exception. The second clear seemes to be more likely correct to do something of the logic `if PyErr_Occurred(): goto fail` to be honest. I am not sure the warnings should be warnings at all and not errors though. Can it happen that we want to warn, then continue and create an object array or so?! We may need a deprecation to turn it to an unconditional error?

File: numpy/core/src/multiarray/ctors.c [hunk 1]
```diff
@@ -1485,14 +1486,16 @@ PyArray_GetArrayParamsFromObject(PyObject *op,
...
-    if (!PyBytes_Check(op) && !PyUnicode_Check(op) &&
-             _array_from_buffer_3118(op, (PyObject **)out_arr) == 0) {
-        if (writeable
-            && PyArray_FailUnlessWriteable(*out_arr, "PEP 3118 buffer") < 0) {
-            Py_DECREF(*out_arr);
-            return -1;
+    if (!PyBytes_Check(op) && !PyUnicode_Check(op)) {
+        if (_array_from_buffer_3118(op, (PyObject **)out_arr) == 0) {
+            if (writeable &&
+                PyArray_FailUnlessWriteable(*out_arr, "PEP 3118 buffer") < 0) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: This doesn't look safe if `out_arr == NULL`...

**Comment by eric-wieser** [comment]
Looks like `_descriptor_from_pep3118_format` needs a fair bit of work here too...

**Pull request closed without merging**